docs/orleans/implementation/grain-directory.md

```diff
@@ -0,0 +1,72 @@
+- View number mismatches trigger synchronization
+- Requests are automatically retried on view changes
+
+This ensures that all requests are processed by the correct owner of the directory partition.
+
+### Partitioning strategy
+
+The directory is partitioned using a consistent hash ring with ranges being assigned to the active silos in the cluster. Grain identifiers are hashed to find the silo which owns the section of the ring corresponding to its hash.
+
+Each active silo owns a pre-configured number of ranges, defaulting to 30 ranges per silo. This is similar to the scheme used by [Amazon Dynamo](https://www.allthingsdistributed.com/files/amazon-dynamo-sosp2007.pdf) and [Apache Cassandra](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archDataDistributeVnodesUsing.html), where multiple "virtual nodes" (ranges) are created for each node (host).
+
+The size of a partition is determined by the distance between its hash and the hash of the next partition. It is possible for a range to be split among multiple silos during a view change, which adds complexity to the view change procedure since each partition must potentially coordinate with multiple other partitions.
+
+### View change procedure
+
+Directory partitions (implemented in `GrainDirectoryPartition`) use versioned range locks to prevent invalid access to ranges during view changes. Range locks are created during view change and are released when the view change is complete. These locks are analogous to the 'wedges' used in the Virtual Synchrony methodology.
+
+When a view change occurs, a partition can either grow or shrink:
+- If a new silo has joined the cluster, then existing partitions may shrink to make room
+- If a silo has left the cluster, then remaining partitions may grow to take over the orphaned ranges
+
+Directory registrations must be transferred from the old owner to the new owner before requests can be served.
+The transfer process follows these steps:
+1. The previous owner seals the range and creates a snapshot of its directory entries
+2. The new owner requests and applies the snapshot
+3. The new owner begins servicing requests for the range
+4. The previous owner is notified and deletes the snapshot
+
+### Recovery process
+
+When a host crashes without properly handing off its directory partitions, the subsequent partition owners must perform recovery. This involves:
+1. Querying all active silos in the cluster for their grain registrations
+2. Rebuilding the directory state for affected ranges
+3. Ensuring no duplicate grain activations occur
+
+Recovery is also necessary when cluster membership changes happen rapidly. While cluster membership guarantees monotonicity, it's possible for silos to miss intermediate membership views. In such cases:
+- Snapshot transfers are abandoned
+- Recovery is performed instead of normal partition-to-partition handover
+- The system maintains consistency despite missing intermediate states
```

Comment:
```suggestion

- Snapshot transfers are abandoned.
- Recovery is performed instead of normal partition-to-partition handover.
- The system maintains consistency despite missing intermediate states.
```